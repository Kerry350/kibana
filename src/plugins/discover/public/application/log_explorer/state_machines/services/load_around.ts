/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { IEsSearchResponse, ISearchSource, QueryStart } from '@kbn/data-plugin/public';
import { DataView } from '@kbn/data-views-plugin/public';
import { catchError, last, map, Observable, of } from 'rxjs';
import { assign } from 'xstate';
import { fetchEntriesAround } from '../../queries/fetch_entries_around';
import { LogExplorerContext, LogExplorerEvent } from '../data_access_state_machine';

export type LoadAroundEvent =
  | {
      type: 'loadAroundSucceeded';
      before: IEsSearchResponse;
      after: IEsSearchResponse;
    }
  | {
      type: 'loadAroundFailed';
      error: unknown;
    };

export const loadAround = ({
  dataView,
  query,
  searchSource,
}: {
  dataView: DataView;
  query: QueryStart;
  searchSource: ISearchSource;
}) => {
  const boundFetchEntriesAround = fetchEntriesAround({ dataView, query, searchSource });

  return (context: LogExplorerContext, event: LogExplorerEvent): Observable<LoadAroundEvent> => {
    // console.log(searchSource.getSearchRequestBody());
    const { position } = context;

    return boundFetchEntriesAround({
      position,
      sorting: [
        [dataView.timeFieldName, 'asc'],
        ['_doc', 'asc'], // _shard_doc is only available when used inside a PIT
      ],
    }).pipe(
      last(),
      map(({ before, after }) => {
        return {
          type: 'loadAroundSucceeded' as const,
          before,
          after,
        };
      }),
      catchError((err) => {
        return of({
          type: 'loadAroundFailed' as const,
          error: err,
        });
      })
    );
  };
};

// TODO: this should be named more specifically and possibly live somewhere else (model?)
export const updateChunks = assign((context: LogExplorerContext, event: LogExplorerEvent) => {
  if (event.type !== 'loadAroundSucceeded') {
    return context;
  }

  return context;
});
