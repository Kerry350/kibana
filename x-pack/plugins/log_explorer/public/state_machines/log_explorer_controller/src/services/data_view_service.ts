/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { InvokeCreator } from 'xstate';
import { LogExplorerControllerContext, LogExplorerControllerEvent } from '../types';

export const createAndSetDataView =
  (): InvokeCreator<LogExplorerControllerContext, LogExplorerControllerEvent> =>
  async (context) => {
    if (!('discoverStateContainer' in context)) return;
    const { discoverStateContainer } = context;
    await discoverStateContainer.actions.createAndAppendAdHocDataView(
      context.datasetSelection.toDataviewSpec()
    );
  };
