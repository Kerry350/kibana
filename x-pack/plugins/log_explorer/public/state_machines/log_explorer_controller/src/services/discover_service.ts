/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { isEmpty } from 'lodash';
import { actions, InvokeCallback } from 'xstate';
import { LogExplorerControllerContext, LogExplorerControllerEvent } from '../types';

export const subscribeToDiscoverState =
  () =>
  (
    context: LogExplorerControllerContext
  ): InvokeCallback<LogExplorerControllerEvent, LogExplorerControllerEvent> =>
  (send, onEvent) => {
    if (!('discoverStateContainer' in context)) {
      throw new Error('Failed to subscribe to the Discover state: no state container in context.');
    }

    const { appState } = context.discoverStateContainer;

    const subscription = appState.state$.subscribe({
      next: (newAppState) => {
        if (isEmpty(newAppState)) {
          return;
        }

        send({
          type: 'RECEIVE_DISCOVER_APP_STATE',
          appState: newAppState,
        });
      },
    });

    return () => {
      subscription.unsubscribe();
    };
  };

export const updateGridFromDiscoverAppState = actions.assign<
  LogExplorerControllerContext,
  LogExplorerControllerEvent
>((context, event) => {
  if ('appState' in event && event.type === 'RECEIVE_DISCOVER_APP_STATE') {
    return {
      grid: {
        columns:
          event.appState.columns?.map((field) => ({
            field,
            width: event.appState.grid?.columns?.[field].width,
          })) ?? context.grid.columns,
        rows: {
          rowHeight: event.appState.rowHeight ?? context.grid.rows.rowHeight,
          rowsPerPage: event.appState.rowsPerPage ?? context.grid.rows.rowsPerPage,
        },
      },
    };
  }

  return {};
});
