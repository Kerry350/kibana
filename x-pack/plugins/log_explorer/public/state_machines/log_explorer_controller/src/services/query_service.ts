/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import type { QueryStart } from '@kbn/data-plugin/public';
import { InvokeCallback } from 'xstate';
import type { LogExplorerControllerContext, LogExplorerControllerEvent } from '../types';

export const subscribeToQueryState =
  ({ query }: { query: QueryStart }) =>
  (
    context: LogExplorerControllerContext
  ): InvokeCallback<LogExplorerControllerEvent, LogExplorerControllerEvent> =>
  (send, onEvent) => {
    const subscription = query.state$.subscribe({
      next: ({ state: queryState }) => {
        send({
          type: 'RECEIVE_QUERY_STATE',
          queryState,
        });
      },
    });

    return () => {
      subscription.unsubscribe();
    };
  };
