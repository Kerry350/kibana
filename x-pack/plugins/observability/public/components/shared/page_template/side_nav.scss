.observabilitySideBar {
  padding: $euiSizeL;
}

.observabilitySideBar__header {
  padding-bottom: $euiSizeL;
}

.observabilitySideBar__headerIcon {
  @include euiBottomShadowSmall;
  margin-right: $euiSize;
}
