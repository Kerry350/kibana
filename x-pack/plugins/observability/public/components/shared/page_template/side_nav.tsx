/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { EuiSideNav } from '@elastic/eui';
import React from 'react';
import type { Observable } from 'rxjs';
import type { NavigationSection } from '../../../services/navigation_registry';

export function ObservabilitySideNav({}: {
  navigationSections$: Observable<NavigationSection[]>;
}): React.ReactElement | null {
  return <EuiSideNav />;
}
