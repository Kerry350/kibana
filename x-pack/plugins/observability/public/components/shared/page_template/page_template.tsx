/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { EuiPageTemplate, EuiPageTemplateProps, ExclusiveUnion } from '@elastic/eui';
import React from 'react';
import { ObservabilitySideNav, ObservabilitySideNavProps } from './side_nav';
import './side_nav.scss';

export type WrappedPageTemplateProps = Pick<
  EuiPageTemplateProps,
  | 'children'
  | 'data-test-subj'
  | 'paddingSize'
  | 'pageBodyProps'
  | 'pageContentBodyProps'
  | 'pageContentProps'
  | 'pageHeader'
  | 'restrictWidth'
> &
  // recreate the exclusivity of bottomBar-related props
  ExclusiveUnion<
    { template?: 'default' } & Pick<EuiPageTemplateProps, 'bottomBar' | 'bottomBarProps'>,
    { template: EuiPageTemplateProps['template'] }
  >;

export type ObservabilityPageTemplateProps = ObservabilitySideNavProps & WrappedPageTemplateProps;

export function ObservabilityPageTemplate({
  children,
  currentAppId$,
  getUrlForApp,
  navigateToApp,
  navigationSections$,
  ...pageTemplateProps
}: ObservabilityPageTemplateProps): React.ReactElement | null {
  return (
    <EuiPageTemplate
      restrictWidth={false}
      {...pageTemplateProps}
      pageSideBar={
        <ObservabilitySideNav
          currentAppId$={currentAppId$}
          getUrlForApp={getUrlForApp}
          navigateToApp={navigateToApp}
          navigationSections$={navigationSections$}
        />
      }
      pageSideBarProps={{
        className: 'observabilitySideBar',
        sticky: true,
      }}
    >
      {children}
    </EuiPageTemplate>
  );
}

// for lazy import
// eslint-disable-next-line import/no-default-export
export default ObservabilityPageTemplate;
