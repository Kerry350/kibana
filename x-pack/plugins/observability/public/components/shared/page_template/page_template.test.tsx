/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React from 'react';
import { shallow } from 'enzyme';
import { createLazyObservabilityPageTemplate } from './lazy_page_template';
import { ObservabilityPageTemplate } from './page_template';
import { createNavigationRegistry } from '../../../services/navigation_registry';
import { ObservabilitySideNav } from './side_nav';
import { of } from 'rxjs';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '/test-path',
  }),
}));

const navigationRegistry = createNavigationRegistry();

navigationRegistry.registerSections(
  of([
    {
      label: 'Test A',
      sortKey: 100,
      entries: [
        { label: 'Url A', app: 'TestA', path: '/url-a' },
        { label: 'Url B', app: 'TestA', path: '/url-b' },
      ],
    },
    {
      label: 'Test B',
      sortKey: 200,
      entries: [
        { label: 'Url A', app: 'TestB', path: '/url-a' },
        { label: 'Url B', app: 'TestB', path: '/url-b' },
      ],
    },
  ])
);

describe('Page template', () => {
  it('Provides a working lazy wrapper', () => {
    const LazyObservabilityPageTemplate = createLazyObservabilityPageTemplate({
      currentAppId$: of('Test app ID'),
      getUrlForApp: () => '/test-url',
      navigateToApp: async (appId) => {},
      navigationSections$: navigationRegistry.sections$,
    });

    const component = shallow(
      <LazyObservabilityPageTemplate
        pageHeader={{
          pageTitle: 'Test title',
          rightSideItems: [<span>Test side item</span>],
        }}
      >
        <div>Test structure</div>
      </LazyObservabilityPageTemplate>
    );
    expect(component).toMatchSnapshot();
  });

  it('Utilises the KibanaPageTemplate for rendering', () => {
    const component = shallow(
      <ObservabilityPageTemplate
        currentAppId$={of('Test app ID')}
        getUrlForApp={() => '/test-url'}
        navigateToApp={async (appId) => {}}
        navigationSections$={navigationRegistry.sections$}
        pageHeader={{
          pageTitle: 'Test title',
          rightSideItems: [<span>Test side item</span>],
        }}
      >
        <div>Test structure</div>
      </ObservabilityPageTemplate>
    );
    expect(component).toMatchSnapshot();
  });

  it('Handles outputting the registered navigation structures within a side nav', () => {
    // TODO: Try to improve this. useObservable inside this component will always use a [] initial value, so the first render
    // (and test result) isn't very useful.
    const component = shallow(
      <ObservabilitySideNav
        currentAppId$={of('Test app ID')}
        getUrlForApp={() => '/test-url'}
        navigateToApp={async (appId) => {}}
        navigationSections$={navigationRegistry.sections$}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
