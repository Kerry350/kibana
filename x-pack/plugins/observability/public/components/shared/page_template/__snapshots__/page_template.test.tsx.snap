// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page template Handles outputting the registered navigation structures within a side nav 1`] = `
<Fragment>
  <EuiTitle
    className="observabilitySideBar__header"
    size="xs"
  >
    <h2>
      <EuiAvatar
        className="observabilitySideBar__headerIcon"
        color="plain"
        iconColor={null}
        iconType="logoObservability"
        name="Observability"
      />
      Observability
    </h2>
  </EuiTitle>
  <EuiSideNav
    items={Array []}
  />
</Fragment>
`;

exports[`Page template Provides a working lazy wrapper 1`] = `
<Suspense
  fallback={null}
  pageHeader={
    Object {
      "pageTitle": "Test title",
      "rightSideItems": Array [
        <span>
          Test side item
        </span>,
      ],
    }
  }
>
  <lazy
    currentAppId$={
      Observable {
        "_isScalar": false,
        "_subscribe": [Function],
      }
    }
    getUrlForApp={[Function]}
    navigateToApp={[Function]}
    navigationSections$={
      Observable {
        "_isScalar": false,
        "operator": [Function],
        "source": Observable {
          "_isScalar": false,
          "operator": MergeMapOperator {
            "concurrent": 1,
            "project": [Function],
          },
          "source": Observable {
            "_isScalar": false,
            "_subscribe": [Function],
          },
        },
      }
    }
    pageHeader={
      Object {
        "pageTitle": "Test title",
        "rightSideItems": Array [
          <span>
            Test side item
          </span>,
        ],
      }
    }
  >
    <div>
      Test structure
    </div>
  </lazy>
</Suspense>
`;

exports[`Page template Utilises the KibanaPageTemplate for rendering 1`] = `
<KibanaPageTemplate
  pageHeader={
    Object {
      "pageTitle": "Test title",
      "rightSideItems": Array [
        <span>
          Test side item
        </span>,
      ],
    }
  }
  pageSideBar={
    <ObservabilitySideNav
      currentAppId$={
        Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        }
      }
      getUrlForApp={[Function]}
      navigateToApp={[Function]}
      navigationSections$={
        Observable {
          "_isScalar": false,
          "operator": [Function],
          "source": Observable {
            "_isScalar": false,
            "operator": MergeMapOperator {
              "concurrent": 1,
              "project": [Function],
            },
            "source": Observable {
              "_isScalar": false,
              "_subscribe": [Function],
            },
          },
        }
      }
    />
  }
  pageSideBarProps={
    Object {
      "className": "observabilitySideBar",
      "sticky": true,
    }
  }
  restrictWidth={false}
>
  <div>
    Test structure
  </div>
</KibanaPageTemplate>
`;
